### PR TITLE
Optimize AppHost startup launch

### PR DESCRIPTION
**Pull request by davidfowl**
## Description

Optimizes .NET project AppHost startup by avoiding the `dotnet run --no-build --project` command path after the CLI has already built the AppHost. The CLI now caches the SDK-provided `RunCommand`, `RunArguments`, and `RunWorkingDirectory` during AppHost resolution and uses that contract to launch the generated apphost executable directly when it is safe to do so.

The direct launch path still preserves the important `dotnet run` behavior: it applies project launch settings through the shared AOT-friendly launch profile reader, keeps launch profile argument precedence aligned with the SDK, handles `dotnet exec`-style run commands, and preserves the repo-local SDK runtime environment for native apphost executables.

### Startup timeline

The measured win is removing the post-build `dotnet run --no-build --project` command path before the AppHost process starts, saving about 1.3s in the profiled TestShop startup trace.

```text
Time after build completes
0ms                 500ms               1000ms              1300ms
|-------------------|-------------------|-------------------|

Before
[ dotnet run --no-build --project overhead .................. ][ AppHost starts ]
  SDK command startup
  launchSettings discovery
  project/run metadata evaluation
  command construction

After
[ cached RunCommand + direct launch ][ AppHost starts ]
  read cached SDK run metadata
  apply launchSettings-compatible env/args
  start generated AppHost command

Saved on AppHost process start: ~1.3s
```

### User-facing usage

Users can opt out of direct AppHost launch if their project relies on custom run behavior that needs the SDK command path:

```bash
aspire config set dotnetAppHostDirectLaunchDisabled true
```

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No

### CODE REVIEW

**Comment by IEvangelist** [comment]
What are the startup times for something like eShop (I'd even consider including several other samples too, starter templates, larger apps, huge apps) today, versus with these changes? You should include some benchmarks maybe?

**Comment by davidfowl** [comment]
> What are the startup times for something like eShop (I'd even consider including several other samples too, starter templates, larger apps, huge apps) today, versus with these changes? You should include some benchmarks maybe?

C# apphost startup time is dominated by the build. The even bigger cost is having project references to the service projects. This is something we will tackle as part of https://github.com/microsoft/aspire/issues/16386.

This fixes a gap between dotnet run and the DAB starting up.

**Comment by davidfowl** [comment]
@copilot resolve the merge conflicts in this pull request

File: src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+
+namespace Aspire.Hosting;
+
+internal static class LaunchSettingsJsonReader
```
Comment (eerhardt) [inline src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs]: What's the difference between this and LaunchSettingsReader?

**Comment by baronfel** [comment]
For full compat with `run`, the apphost's Run* properties should be cached after running the `ComputeRunArguments` target. If you're using the command-line `-getproperty` etc arguments, you can add the target via `-t:ComputeRunArguments`. This will have MSbuild run the target and then give you the same requested properties/items/etc, just after the Target has run.

**Comment by davidfowl** [comment]
Thanks for the overview. This was informational, so no code change was needed for this summary comment.

File: src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs (hunk 1 above)
Comment (davidfowl) [reply src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs]: LaunchSettingsJsonReader is the low-level generic reader: it takes a caller-supplied JsonTypeInfo<T> and surfaces JSON parse failures as InvalidDataException, so callers outside Aspire.Hosting can read launchSettings.json without depending on the concrete LaunchSettings shape or DistributedApplicationException. LaunchSettingsReader is now a thin wrapper around it for the Aspire.Hosting LaunchSettings type, and I added a class-level XML doc to make that distinction explicit.

File: src/Aspire.Cli/Caching/AppHostInfoDiskCache.cs [hunk 2]
```diff
@@ -129,6 +128,15 @@ public AppHostInfoDiskCache(ILogger<AppHostInfoDiskCache> logger, CliExecutionCo
                 return null;
             }
 
+            if (entry.ExitCode == 0 && entry.IsAspireHost && string.IsNullOrWhiteSpace(entry.RunCommand))
+            {
+                // v1 has not shipped yet, so we keep the schema marker stable. Local pre-merge cache
+                // entries created before RunCommand was added should be treated as misses so the direct
+                // launch path can refresh the SDK run command instead of permanently falling back.
+                _logger.LogTrace("AppHost info cache entry for {Project} is missing RunCommand", projectFile.FullName);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Caching/AppHostInfoDiskCache.cs:137]: Good catch. I reworded the comment to describe the durable behavior: entries that successfully inspected an AppHost but are missing RunCommand are treated as schema-incompatible misses so the direct launch path can refresh the SDK run command.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 3]
```diff
@@ -512,6 +535,328 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+                hasExplicitApplicationArgs: unmatchedTokens.Length > 0,
+                hasRunArguments))
+        {
+            return null;
+        }
+
+        foreach (var (name, value) in env)
+        {
+            directEnv[name] = value;
+        }
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Projects/DotNetAppHostProject.cs:582]: Intentional: env here is the CLI-curated environment that carries Aspire control variables like isolated-mode ports, AspireLogLevel, WaitForDebugger, dev-cert settings, and CLI bundle config, so those values must win even if a launch profile uses the same key. The existing direct-launch test asserts this precedence, so I left the behavior unchanged.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 4]
```diff
@@ -512,6 +535,328 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+            profileName = null!;
+            profile = null!;
+            return false;
+        }
+
+        foreach (var (candidateProfileName, candidateProfile) in launchSettings.Profiles)
+        {
+            if (IsSupportedLaunchProfile(candidateProfile))
+            {
+                profileName = candidateProfileName;
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Cli/Projects/DotNetAppHostProject.cs:858]: Good catch. I added null guards for a null Profiles map, null profile entries, null EnvironmentVariables, and null environment values, with focused tests covering malformed-but-parseable launchSettings so direct launch falls back or skips null entries instead of throwing.

File: src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs (hunk 1 above)
Comment (eerhardt) [reply src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs]: I still don't understand. These 2 classes look like they do the same thing.

File: src/Aspire.Cli/Projects/AppHostLaunchSettings.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Aspire.Cli.Projects;
+
+internal sealed class AppHostLaunchSettings
+{
+    [JsonPropertyName("profiles")]
+    public Dictionary<string, AppHostLaunchProfile> Profiles { get; set; } = [];
+}
+
+internal sealed class AppHostLaunchProfile
```
Comment (eerhardt) [inline src/Aspire.Cli/Projects/AppHostLaunchSettings.cs:15]: It's unfortunate that this is duplicated with Aspire.Hosting.LaunchProfile.

File: src/Aspire.Cli/Projects/AppHostInfoResolver.cs [hunk 6]
```diff
@@ -223,7 +242,13 @@ internal sealed record AppHostProjectInfo(
...
     string? AspireHostingVersion,
     bool IsUsingCliBundle,
-    string? UserSecretsId);
+    string? UserSecretsId,
+    string? RunCommand,
+    string? TargetPath,
+    string? RunWorkingDirectory,
+    string? RunArguments,
+    string? TargetFramework,
+    string? TargetFrameworks);
```
Comment (eerhardt) [inline src/Aspire.Cli/Projects/AppHostInfoResolver.cs:265]: What do we use TargetFrameworks for?

File: src/Aspire.Hosting.Testing/Aspire.Hosting.Testing.csproj [hunk 7]
```diff
@@ -13,6 +13,7 @@
 
   <ItemGroup>
     <Compile Include="$(SharedDir)LaunchProfiles\LaunchSettingsReader.cs" Link="LaunchSettingsReader.cs" />
+    <Compile Include="$(SharedDir)LaunchProfiles\LaunchSettingsJsonReader.cs" Link="LaunchSettingsJsonReader.cs" />
```
Comment (eerhardt) [inline src/Aspire.Hosting.Testing/Aspire.Hosting.Testing.csproj]: Why this change?

**Comment by eerhardt** [comment]
> For full compat with `run`, the apphost's Run* properties should be cached after running the `ComputeRunArguments` target. If you're using the command-line `-getproperty` etc arguments, you can add the target via `-t:ComputeRunArguments`. This will have MSbuild run the target and then give you the same requested properties/items/etc, just after the Target has run.

I don't see this comment being addressed. Was it?

File: src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs (hunk 1 above)
Comment (davidfowl) [reply src/Shared/LaunchProfiles/LaunchSettingsJsonReader.cs]: This split is unnecessary. This happened when I mentioned the AOT split

**Comment by davidfowl** [comment]
Addressed in 524ab97000f362a1b60fc2877cd357f4942e8a18. `AppHostInfoResolver` now passes `targets: ["ComputeRunArguments"]` to `GetProjectItemsAndPropertiesAsync`, and `DotNetCliRunner` emits `-t:ComputeRunArguments` so MSBuild runs the target before evaluating `-getProperty` for `RunCommand`/`RunArguments`/`RunWorkingDirectory`/`TargetPath`. Added tests for both the runner switch generation and the resolver wiring.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 8]
```diff
@@ -326,6 +332,10 @@ public async Task<int> RunAsync(AppHostProjectContext context, CancellationToken
             ConfigureSingleFileRunEnvironment(effectiveAppHostFile, env, args: context.UnmatchedTokens);
         }
 
+        var directRun = !isSingleFileAppHost && !watch && !isExtensionHost
+            ? await TryCreateDirectRunSpecAsync(effectiveAppHostFile, env, context.UnmatchedTokens, runOptions.NoLaunchProfile, cancellationToken)
```
Comment (JamesNK) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs:336]: **[Automated review]** Potential correctness issue — stale cache can launch an outdated executable after rebuild.

`TryCreateDirectRunSpecAsync` reads `RunCommand` from the disk cache, which was populated during a previous resolution cycle. If the user changes a property that affects the output path (e.g. `OutputPath`, `AssemblyName`, or a build-time conditional), the cache may return the old path. The `File.Exists` check in `TryResolveDirectRunTarget` catches the case where the old file is gone, but if the old binary still exists on disk (e.g. switching between `Debug`/`Release` configurations), the CLI will launch the stale binary instead of the freshly built one.

The fallback to `dotnet run` would correctly resolve the current `RunCommand` via MSBuild's `ComputeRunArguments` at launch time.

This should either be addressed (e.g. by invalidating the cached `RunCommand` when the build output path changes, or by verifying the cached path against `TargetPath` from the current build), or a comment should be added explaining why this scenario is acceptable (e.g. cache fingerprint already covers these inputs).

File: src/Aspire.Cli/Projects/AppHostInfoResolver.cs [hunk 9]
```diff
@@ -103,21 +103,37 @@ private async Task<AppHostProjectInfo> FetchAppHostInfoCoreAsync(FileInfo projec
...
+        // resolve before launching the AppHost. TargetFrameworks is also cached so
+        // DotNetAppHostProject can detect multi-targeted AppHosts and fall back to
+        // `dotnet run`, which already handles target framework selection. See:
+        // https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
         var (exitCode, jsonDocument) = await runner.GetProjectItemsAndPropertiesAsync(
             projectFile,
             items: ["PackageReference", "AspireProjectOrPackageReference", "PackageVersion"],
-            properties: ["IsAspireHost", "AspireHostingSDKVersion", "AspireUseCliBundle", "UserSecretsId"],
+            properties: ["IsAspireHost", "AspireHostingSDKVersion", "AspireUseCliBundle", "UserSecretsId", "RunCommand", "TargetPath", "RunWorkingDirectory", "RunArguments", "TargetFramework", "TargetFrameworks"],
+            targets: ["ComputeRunArguments"],
```
Comment (JamesNK) [inline src/Aspire.Cli/Projects/AppHostInfoResolver.cs:141]: **[Automated review]** Performance concern — passing `targets: ["ComputeRunArguments"]` makes the AppHost info probe execute targets rather than just evaluating properties.

`ComputeRunArguments` likely has dependencies on the `Build` target (it reads `TargetPath`). When this probe runs in non-`run` code paths (e.g. `aspire ls`, project discovery, or cache-miss scenarios before the CLI has built the project), it may:
1. Trigger a full build as a side effect (if MSBuild resolves the target dependency chain), or
2. Return empty/incorrect `RunCommand` values because the build outputs don't exist yet.

Previously this was a lightweight evaluation-only call. Now every cache miss in any code path that calls `GetAppHostInfoAsync` pays the cost of target execution.

This should either be addressed (e.g. only requesting the `ComputeRunArguments` target when the caller is on the `run` path and a build has already completed), or a comment should be added explaining why this is acceptable (e.g. `ComputeRunArguments` is always cheap/self-contained and doesn't trigger compilation).

**Review by JamesNK — COMMENTED** [review]
Automated review — 2 findings (1 correctness, 1 performance). These should either be addressed, or a comment added to the code explaining why the current behavior is correct.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 10]
```diff
@@ -512,6 +535,348 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+        return true;
+    }
+
+    private static bool HasMultipleTargetFrameworks(AppHostProjectInfo info)
+        => info.TargetFrameworks?.Split(';', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Length > 1;
+
+    private static string ResolvePath(string path, DirectoryInfo baseDirectory)
+        => Path.IsPathFullyQualified(path) ? path : Path.GetFullPath(Path.Combine(baseDirectory.FullName, path));
+
+    private static string NormalizeCommand(string command)
```
Comment (JamesNK) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs]: **[Automated review]** Minor: `NormalizeCommand` is now duplicated in three places with identical logic — here, in `DotNetCliRunner.cs` (line 194), and in `CommandPathResolver.cs` (line 81). If this diverges in the future it could cause subtle behavioral differences between the direct-launch path and the runner's muxer detection.

Consider extracting to a shared helper, or add a comment noting the intentional duplication and that the copies must stay in sync. Otherwise this can be left as-is if the duplication is acceptable for the scope of this PR.

**Review by JamesNK — COMMENTED** [review]
Automated review — 1 additional minor finding (code duplication).

**Review by eerhardt — APPROVED** [review]

**Review by radical — COMMENTED** [review]
Two follow-ups on the direct-launch path that aren't blockers, but would help the next maintainer who hits a stale-cache or contract-regression symptom in the wild.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 11]
```diff
@@ -512,6 +535,352 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+        }
+
+        foreach (var (name, value) in env)
+        {
+            directEnv[name] = value;
+        }
+
+        arguments.AddRange(unmatchedTokens);
+
+        return new DirectAppHostRunSpec(command, workingDirectory, [.. arguments], directEnv);
```
Comment (radical) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs:592]: Direct launch has `LogDebug` on every fallback branch but no log on the success path. If a user reports "Aspire is launching something that looks like an old build" (the bounded stale-cache scenario for out-of-tree imports the cache doc already calls out), `--debug` gives them no breadcrumb that direct launch was chosen or which cached `RunCommand` was used — every visible log message is about falling back to `dotnet run`. Consider a `_logger.LogDebug("Launching AppHost directly via {Command} {WorkingDirectory} with arguments {Arguments}.", command, workingDirectory.FullName, arguments)` here before the `return`, so the success path is symmetric with the fallbacks and the cache-staleness recovery path documented on `AppHostInfoDiskCache` is actually discoverable from a debug log.

File: tests/Aspire.Cli.Tests/DotNet/DotNetCliRunnerTests.cs [hunk 12]
```diff
@@ -1621,6 +1705,82 @@ await runner.GetProjectItemsAndPropertiesAsync(
...
+            projectFile,
+            items: [],
+            properties: ["RunCommand"],
+            targets: ["ComputeRunArguments", "Build"],
+            options,
+            CancellationToken.None
+        );
+
+        Assert.Equal(0, exitCode);
+        Assert.NotNull(output);
```
Comment (radical) [inline tests/Aspire.Cli.Tests/DotNet/DotNetCliRunnerTests.cs:1762]: This test (and its sibling at line 1754) verifies command-line shape with a hard-coded `RunCommand` injected into mocked stdout. There's no test anywhere that proves a real `dotnet msbuild -t:ComputeRunArguments;Build -getProperty:RunCommand,RunArguments,RunWorkingDirectory` actually populates those properties for the kind of project the CLI builds.

The `Aspire.Cli.EndToEnd.Tests/SmokeTests.cs` `aspire new` → `aspire run` flow does exercise a real SDK, but if `ComputeRunArguments` ever stops setting `RunCommand` (SDK change, target rename, etc.), `TryResolveDirectRunTarget` falls back via `string.IsNullOrWhiteSpace(info.RunCommand)` and the user silently regresses to the slow `dotnet run --no-build --project` path — SmokeTests still passes, and the whole perf win this PR is buying disappears unnoticed.

Consider one of:
- A positive E2E assertion that direct launch was used (e.g. wait for the success-path debug log line suggested in the other comment, or assert absence of an `aspire run`-only signature).
- A targeted integration test that runs real `dotnet msbuild` against a minimal temp project and asserts `RunCommand` / `RunArguments` are populated.

Either catches a future SDK contract drift loudly instead of silently.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 13]
```diff
@@ -512,6 +535,352 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+                    runtimeConfigPath);
+                return false;
+            }
+
+            command = runCommand;
+        }
+        else
+        {
+            var resolvedRunCommand = ResolvePath(runCommand, projectDirectory);
+            if (!File.Exists(resolvedRunCommand))
```
Comment (radical) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs:674]: **Safety-net fallback tests are missing.** The three `File.Exists` guards that make direct-launch safe against a stale cache — line 644 (`dotnet exec` target), line 654 (`runtimeconfig.json`), line 668 (native `RunCommand`) — have no positive tests. Every direct-launch test creates the target file via `CreateBuiltAppHostAssembly` / `CreateBuiltAppHostCommand` (which also creates `.runtimeconfig.json` at line 1259), so the "build output is gone, fall back to `dotnet run`" path is never exercised. These are the exact guards the cache-staleness rationale at lines 551–557 relies on for correctness.

Recommend three small tests that withhold each file in turn and assert `RunAsync` falls through to the SDK path. While there, the same shape covers a few other trivially missing cases that all fail closed and have no test: `HasMultipleTargetFrameworks` (line 606), `RunCommand` whitespace/missing (line 615), and `RunCommand=dotnet` but `RunArguments` not starting with `exec` (line 635).

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 14]
```diff
@@ -326,6 +332,10 @@ public async Task<int> RunAsync(AppHostProjectContext context, CancellationToken
             ConfigureSingleFileRunEnvironment(effectiveAppHostFile, env, args: context.UnmatchedTokens);
         }
 
+        var directRun = !isSingleFileAppHost && !watch && !isExtensionHost
```
Comment (radical) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs:335]: **`watch` and `isExtensionHost` gates aren't pinned by tests.** Direct launch is gated by `!isSingleFileAppHost && !watch && !isExtensionHost`. The single-file branch is covered implicitly by existing tests; the other two are not. In particular there's no assertion that `aspire watch` continues to take the SDK `dotnet run` path (required for the rebuild-on-save flow) — a future refactor could accidentally enable direct launch in watch mode and silently break it. Two small tests asserting the SDK fallback for `watch:true` and for an extension-host scenario would pin the contract.

File: src/Aspire.Cli/Projects/DotNetAppHostProject.cs [hunk 15]
```diff
@@ -512,6 +535,352 @@ private async Task EnsureDevCertificatesTrustedAsync(AppHostProjectContext conte
...
+                foreach (var (name, value) in profile.EnvironmentVariables)
+                {
+                    if (value is null)
+                    {
+                        // `System.Text.Json` will deserialize `"FOO": null` into a null dictionary
+                        // value even though the value type is non-nullable. Skip those entries.
+                        continue;
+                    }
+
+                    env[name] = Environment.ExpandEnvironmentVariables(value);
```
Comment (radical) [inline src/Aspire.Cli/Projects/DotNetAppHostProject.cs:773]: **`%VAR%` expansion is untested and deviates from `dotnet run`.** Both env values (line 765) and `commandLineArgs` (line 771) are passed through `Environment.ExpandEnvironmentVariables`. The SDK's `dotnet run` does NOT expand `%VAR%` in launch-profile env values — `RunCommand.cs` `SetEnvironmentVariables` calls `command.EnvironmentVariable(entry.Key, entry.Value)` with no expansion. Direct launch is matching `Aspire.Hosting.ProjectResourceBuilderExtensions.cs:708` (which expands for child project resources) instead of matching the SDK behavior the PR description claims to preserve.

This is probably the right call for consistency inside Aspire, but it's worth (a) a test pinning the expansion on a representative value like `"%USERPROFILE%\\.myapp"` so a future refactor can't silently drop the call, and (b) a one-line comment at line 765 noting this is intentionally Aspire-aligned, not SDK-aligned, so future maintainers don't "fix" it back to SDK semantics.

**Review by radical — COMMENTED** [review]
Test-coverage follow-ups from auditing the new direct-launch branches against the test suite. Three gaps where the code has explicit defensive paths or pipeline gates with no positive test pinning the behavior.
